### ISSUE TITLE
Infinitely looping queue

### ISSUE THREAD

**Issue by gopherclass** [gocolly/colly#450]
Current queue implementation loops infinitely if queue is empty and active.

https://github.com/gocolly/colly/blob/b1a8ed2f18144f4b70abcfc18a5e58c68a062389/queue/queue.go#L125-L132

![pprof](https://user-images.githubusercontent.com/56159921/75090997-f962ed00-55ab-11ea-8a8e-89c06391dd9b.png)

**Comment by tobinski** [comment]
I think to fix this we need to change the thread goroutine to something like this.
```go
go func() {
	defer wg.Done()
	for r := range q.requestsOut {
                //  this needs to be inside the for loop
		atomic.AddInt32(&q.activeThreadCount, 1)
		r.Do()
		atomic.AddInt32(&q.activeThreadCount, -1)
	}
}()
```

**Comment by gopherclass** [comment]
The following scenario is possible. Queue has one request which takes
longer to complete (e.g. httpbin/delay/10). First, the producer goroutine
sends the request to the consumer goroutine. The consumer executes the
request. Returning to the producer, the producer enters if block since the
queue is empty. q.activeThreadCount is 1. Nested if is not executed and
continue statement is reached. This process is repeated until the request
is completed (for 10 seconds). Your scenario cannot handle the scenario.

My opinion is that we change Storage.GetRequest to be blocking, or use
Queue as Storage when Storage is passed as a argument to queue. I think
either can be a breaking change.



2020년 2월 26일 (수) 오전 2:31, tobinski <notifications@github.com>님이 작성:

**Comment by gopherclass** [comment]
I opened PR #452 to address this issue.

**Comment by asciimoo** [comment]
hopefully fixed by #452
